Min # Vending step2

undervineg opened the pull request.

undervine committed:
Tide up files related to MVC's model into 1 folder

undervine committed:
Separate README.md contents into several .md files

undervine committed:
Add links and modify image path

undervine committed:
Add image views, buttons and labels in StackViews using AutoLayout

undervine committed:
Post notification when stock or balance updated

undervine committed:
Add Enum that defines Notification Names which are used in NotificationCenter-

undervine committed:
Add currency() that makes Int to String in decimal format

undervine committed:
Add <Self> to the parameter of insertMoney()

undervine committed:
Modify some comments

undervine committed:
Add IBActions, IBOutlets, and functions to manipulate model-the vending machine

undervine committed:
Add views

undervine committed:
Add contents about step2

undervine committed:
Add working hours

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -18,12 +21,66 @@ class ViewController: UIViewController {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(updateBalanceLabel),
+            name: Notification.Name(NotificationNames.didUpdateBalance.description),
```
`Notification.Name(NotificationNames.didUpdateBalance.description)` 형태보다는 Notification.Name() 자체를 여러개 만들어 놓고 활용하는게 더 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -18,12 +21,66 @@ class ViewController: UIViewController {
+    // 재고 추가 버튼 클릭 시. V -> C -> M
+    @IBAction func addStock(_ sender: UIButton) {
+        // 이벤트가 발생한 버튼
+        for button in addStockButtons where button.tag == sender.tag {
```
sender가 이미 button 인데 굳이 배열에서 찾을 필요가 있나요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -18,12 +21,66 @@ class ViewController: UIViewController {
+    // 금액 추가 버튼 클릭 시. V -> C -> M
+    @IBAction func insertMoney(_ sender: UIButton) {
+        // 버튼 태그에 따라 특정 금액 삽입.
+        machine.insertMoney(MoneyManager<VendingMachine>.Unit(rawValue: sender.tag)!)
```
화면 요소와 데이터를 매칭해야 하는 경우는 항상 발생합니다.
이렇게 직접적으로 스토리보드에서 태그값을 넣어놓고 sender.tag 에 접근하기 보다는 화면요소와 데이터를 매칭할 수 있는 데이터 구조를 갖고 있는게 좋습니다.
값을 바꾸거나 데이터 구조에 따라서 화면 요소까지 바꿔야 하기 때문에 분리하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Utilities/Extensions.swift:
```diff
@@ -39,3 +39,34 @@ extension EnumCollection {
+        formatter.numberStyle = .decimal
+        return formatter.string(from: self as NSNumber)
+   }
+}
```
이렇게 하나의 파일로 extension을 합치면 어떤 장단점이 있을까요?
귀찮더라도 익스텐션을 분리해놓는게 필요한 경우에만 선택적으로 확장할 수 있는 장점이 있습니다. 여기서는 크게 영향을 미치지 않지만, 일반적인 타입을 확장할 경우에는 특히 더 주의할 필요가 있습니다.

undervineg replied:
음.. 각 익스텐션별로 파일을 따로 만드는 게 좋다는 말씀이신가요~?

godrm replied:
지금은 크게 영향을 미치지 않을 수 있고, 파일을 나누는 게 필요한 경우도 있다는 겁니다 :)

godrm commented on VendingMachineApp/VendingMachineApp/Model/Utilities/Extensions.swift:
```diff
@@ -39,3 +39,34 @@ extension EnumCollection {
+        newDictionary.updateValue(newValue, forKey: key)
+        return newDictionary
+    }
+}
```
(비슷한 피드백을 준 적이 있었던 것 같은데...) 콜랙션 객체를 mutating 함수로 확장하는 것보다는 새로운 객체를 리턴하도록 작성하는 편이 좋습니다. 이런 구조를 선택한다면 어떤 기준에서 왜 꼭 필요한 거 설득이 필요합니다.

undervineg replied:
아 고쳐놓고 mutating 키워드를 지우는 걸 깜빡했네요 ;;

godrm commented on VendingMachineApp/VendingMachineApp/Model/MoneyManager.swift:
```diff
@@ -9,12 +9,16 @@
+            NotificationCenter.default.post(
+                name: NSNotification.Name(NotificationNames.didUpdateBalance.description),
+                object: nil)
+        }
```
Observer 패턴을 벌써 적용했군요! 이거와 관련된 부분은 그 단계에서 피드백줄게요.

godrm reviewed (commented):
아래 항목들을 확인해보세요.
그리고 지금은 스토리보드보다 코드로 객체를 만들고 표현하는 연습을 좀 더 하세요. layer.cornerRadius 값을 디자인 타임에 적용해서 확인하는 것도 좋지만, 코드로 속성을 어디에 넣어야 바뀌는지 공부하고 동작 방식을 이해하는 게 더 중요합니다.

undervine committed:
Move 'layer.cornerRadius' from storyboard to ViewController

Remove IBActions and use addTarget() to set events to UI objects.

undervine committed:
Remove 'mutating' keyword from update() of Dictionary extension

undervine committed:
Remove 'for' statement

Remove 'for' statement searching which button is touched, because addStock() function gets sender as parameter.

undervine committed:
Use a computed property in enum 'Notifications' which returns Notification.Name type value

Remove 'CustomStringConvertible' from the Notifications enum.

undervine committed:
Change StackView's x, y coordinates

undervine committed:
Add Mapper struct to mapping view with model

Add mapping- functions for each UI element.

undervine committed:
Add contents about PR comments

undervineg commented:
- layer.cornerRadius를 코드로 작성
- Dictionary extension의 update() 함수의 mutating 키워드 제거
- IBAction 모두 제거, 코드에서 addTarget을 통해 UI에 이벤트 연결
- Notifications 열거형에 연산프로퍼티를 사용하여 Notification.Name 타입 값을 리턴하도록 변경
- Mapper 구조체를 생성하여 view와 model을 매핑하는 함수 추가

godrm commented on VendingMachineApp/VendingMachineApp/Mapper.swift:
```diff
@@ -0,0 +1,58 @@
+        default: break
+        }
+        return menu
+    }
```
`mappingMenu(with sender: UIButton)` 와 `mappingMenu(with sender: UILabel)` 도 코드가 중북된 것 같지 않나요?
- tag속성은 UIView에 있습니다. 합칠 수 있겠죠? 
- Menu가 enum 이면 switch 로 비교하기 보다는 enum.init() 생성함수로 매칭하면 어떨까요? 입력하면서도 귀찮다고 느끼지 않았나요?

undervineg replied:
enum.init()라면.. Menu 열거형에 rawValue 타입을 주고 init(rawValue:)를 사용하라는 말씀이신가요~?

godrm replied:
네 어차피 enum 타입을 만드는거니까요.

undervineg replied:
그러면 Menu 열거형에 의미없는(?) Int형 rawValue를 주고, init(rawValue:)할 때 인자값으로 tag 값이 들어가게 될텐데 .. 결국에는 tag 값을 직접 쓰는 거나 마찬가지이지 않을까 싶어서요.

godrm replied:
init?(rawValue:) 를 쓰면 sender.tag 를 넣더라도 범위에 들어가면 열거형 타입이 생성될꺼고 아니면 nil 값이 됩니다. 
결과적으로는 동일하지만 switch-case 로 모든 값을 매칭해주지 않아도 init() 함수로 매칭할 수 있는거죠. 
이렇게 작성할 경우 유지보수나 확장성 측면에서도 Menu에 case가 추가되거나 삭제해도 이 부분을 수정할 필요가 없다는 장점이 있습니다. 개발자가 코드로 실수할수 있는 부분을 컴파일러가 검사할 수 있도록 만드는 방식이 있다면 그걸 선택하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -18,12 +23,74 @@ class ViewController: UIViewController {
+        // 버튼 태그에 따라 특정 금액 삽입.
+        machine.insertMoney(
+            MoneyManager<VendingMachine>.Unit(rawValue: Mapper.mappingUnit(with: sender))!)
+    }
```
mappinpUnit() 함수가 Unit 타입을 리턴하면 어땠을까요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -18,12 +23,74 @@ class ViewController: UIViewController {
+    @objc func updateBalanceLabel() {
+        // 잔액라벨 업데이트.
+        if let balance = machine.showBalance().currency() {
+            balanceLabel.text = balance + "원"
```
이렇게 단위를 붙이는 것도 여러 국가를 지원할 때를 고려해서 하드코딩 값을 넣기보다는 Formatter를 만드는게 좋습니다.

godrm reviewed (commented):
개선한 내용이 좋습니다. 다음 사항을 확인해보세요.

undervine committed:
Remove one of mappingMenu() functions

Modify parameter of the other one to UIView, making the sender available regardless of the type of UI element.

undervine committed:
Modify Mapper.mappingUnit() func to return MoneyManager.Unit type, not Int

undervine committed:
Add Formatter to support multi-language

undervine committed:
Add contents about feedback

undervineg commented:
- Mapper.mappingMenu() 메소드의 파라미터를 UIView로 바꾸어 중복되는 로직의 함수 제거
- Mapper.mappingUnit() 메소드가 MoneyManager.Unit 타입을 리턴하도록 변경
- Formatter 열거형을 만들어 기존의 '원', '개' 하드코딩 제거

undervine committed:
Resize the image

godrm commented:
다른 부분들은 변경한 내용이 좋네요. 위에 enum 관련된 내용만 확인하고 넘어가겠습니다.

undervine committed:
Add an image

undervine committed:
Edit '원' in moneyUnit's .eng case to 'won'

undervine committed:
Make RawValue type 'Int'

undervine committed:
Replace switch statements with Enum.init()

undervine committed:
Modify addBalance buttons' tags to the price each button presents as label text

undervine committed:
Add contents about feedback

godrm commented:
수고하셨습니다. 다음 단계로 갑시다
